### DISCUSSION TITLE
Would a decentralized user verification be possible/useful?

### DISCUSSION THREAD

**Discussion by Andre601** [mastodon/mastodon#20160] [General]
This was a question going through my head for a while now since I started using Mastodon.

## Situation
Mastodon's way of handling users allows anyone to create a new account with the exact same name on a different instance, as names aren't "locked" across the Fediverse itself.
This causes some problems, as people could easily make a new account with the name of a popular person on Mastodon on another instance, post messages to explain the situation (i.e. "I lost access so I moved here" or "This has more features/benefits") and then share scams or similar.

Without verification is there no way to really draw a line between who is real and who not.
And yes, link verification exists, but the current way/implementation is "bad" in a way, as the person could simply get a domain with either different tld or just a typo, copy the original website and verify that to look legit (For more proof that this works, see any Discord scam attempt ever).

With verification, there would be *some* level of trust, as someone would've spend the time to check the account to see if it is legit or fake.

## Possible issues and limitations
The way Mastodon is designed - focussing on Federation - doesn't really allow an easy way of verification to be done here.

You either do a centralized system or a de-centralized system which both have their own issues I cover real quick.

### Centralized system
With a centralized system would the main issue be that one authority would call the shots, which would go against the entire point of Mastodon being de-centralized (Federated) and therefore not be a good option.

### De-centralized system
The only way I can see a de-centralized system work is by having admins of the instance in question do the verification.
This comes with the issue that someone can make their own instance and verify themself, which in the end would make verification less useful as pretty much anyone could imposter someone now.

An idea to "fix" this could be to have other instances not "accept" the verification at first. Maybe have the badge not display or have something like a `?` shown, where hovering over it would tell you that the user was verified by "domain.tld" but the verification hasn't been accepted by this instance.
The admins of that instance could then either accept or deny the verification, resulting in a verified badge or something like an `X` for denied verifications to be displayed...

Here are some mockups I made showing the situations where I would be verified by the instance I'm on (blobfox.coffee) and the example instance example.com I view this from...

- ...has not accepted/denied the verification yet
  ![grafik](https://user-images.githubusercontent.com/11576465/201394965-b1d45af4-88c3-4ac7-9fa0-d84eec63a005.png)
- ...has accepted the verification (Just replace "approved" with "accepted")
  ![grafik](https://user-images.githubusercontent.com/11576465/201395099-d1c91c14-6dcd-4fb5-b2be-23b9e8c88c0a.png)
- ...has denied the verification. In this example would the badge not be displayed.
- ...has denied the verification AND also flagged the user
  ![grafik](https://user-images.githubusercontent.com/11576465/201395960-5269220d-a71f-49a4-9a66-4fb5eca8da3e.png)

In order for an instance to review the verification, a user could perhaps request to be reviewed?

**Comment by Andre601** [comment]
Another thing that came to mind is user names...

From how I understand the system behind Mastodon are claimed usernames not shared across other instances, meaning if I would make an account on `mastodon.online` could someone else still make an account with the same name on another instance.
If that is truly the case is there another (relatively large) risk of having scamers and identity-theft...

The main issues remain tho... It would require a system that is either centralized (bad) or relies on the admins/managers of the instance (Also bad).

It's quite something to think about...

**Comment by darmengol** [comment]
Hi, never participated in a Github discussions, but I feel that this is an interesting topic. I'm an information architect so I would talk only from this point of view.

I think that current link verification on Mastodon is a useful approach, relying on external authority. For example, a journalist can be verified by their main magazine with which they collaborate. It reminds me how digital certificates work, where at the end of the chain you must trust some issuer.

The current problem that I see with this is how its reflected on the UI: unlike the bird site, users can only see this information if they explicitly go to the profile. Simply putting a "check" next to the handle can't work within this model, because the user must acknowledge the "issuer" of authority.

I was thinking that maybe a solution could be to add next to the handle (small, but visible), the verification information, for example:

- "Martin Belam _from theguardian.com/profile/martin-belam_"
- "Mastodon _from joinmastodon.org_"
- "Dani Armengol _from daniarmengol.com_"

**Reply by Andre601** [reply]
The main issue I see with link verification as of now is, that a scammer or alike could make an account with the same name on a separate instance, setup a copy of the website and use a domain that looks/sound simila (see any discord "free nitro scam" ever) to "verify".

People may not spot the difference and due to the site looking the same fall for the trick.

That's why I think link verification, while being made in good faith, is a flawed system.

What I think of right now being a *usable* system would be having verification be done by the instance admins of where the account is hosted on.
Then, whenever a user from a different instance views the account could thee be an icon, informing that "user was verified by example.com".
And for extra safety/confirmation could other instances "accept"/"aknowledge" the verification as being legit (or even deny it)

Just a rough idea I had that could keep it decentralized.

**Reply by darmengol** [reply]
It's an interesting approach, but the extra safety/confirmation would be always required, because anyone could create an instance and verify themselves (following the certification analogy, anyone can self-sign a certificate, this is why some external trust in an issuer is needed).

**Reply by Andre601** [reply]
That's why I mentioned that admins of other instances could approve or deny it, showing clear indicators if said account is trusted or not.

Lime imagine instances `masto1.example.com` and `masto2.example.com` with a user called `user` on the first instance and the following cases:

1. User is not verified on his instance
2. User is verified on his instance, but second instance has not accepted nor denied it yet
3. User is verified on his instance and the second one has accepted it
4. User is verified on his instance but the second one has denied it

In these cases could different badges and/or info be displayed on the user's name:

1. No badge at all
2. Verified badge on masto1, but no badge (or maybe something like a `?` badge) on masto2 with a text like "Verified on Masto1.example.com but not yet approved on Masto2.example.com"
3. Verified badge on both masto1 and masto2 with first saying "Verified" and second saying something like "Verified on Masto1.example.com" or similar
4. Verified badge on masto1, but on masto2, a different badge (i.e. an `X`) is displayed, warning about the verification being not accepted by masto2

**Comment by mitexleo** [comment]
For Username Mastodon needs to implement the Zot Protocol .. Which will then allow the usage of nomadic identity...

**Comment by kohtala** [comment]
Digital identity is or is becoming available for user verification in many parts of the world.

EU has [eIDAS](https://digital-strategy.ec.europa.eu/en/policies/discover-eidas). Distributed in the sense that each country manages identity for their citizens providing the trust, but also [cross-country identity system](https://ec.europa.eu/digital-building-blocks/wikis/display/EIDCOMMUNITY/Overview+of+pre-notified+and+notified+eID+schemes+under+eIDAS) for trust across countries and some common standards. Currently several countries have own national digital identities that are more difficult to use in part because the browser ecosystem is not up-to-date and each country has to be managed differently.

eIDAS is developing Digital Wallet for [digital identity](https://ec.europa.eu/info/strategy/priorities-2019-2024/europe-fit-digital-age/european-digital-identity_en). Based on decentraliced technologies, like [Self-Issued OpenID Provider v2](https://openid.net/specs/openid-connect-self-issued-v2-1_0.html), it is currently being [implemented in Finland](https://dvv.fi/en/European-digital-identity-wallet) (Suomi, in Finnish). Follow the link and you can watch a video - it will work also on-line, by web service giving a QR code and mobile providing data directly to service. There is [a decription of it](https://wiki.dvv.fi/display/DHHJD/SIOPv2+step-by-step+process+and+relating+trust+model) and there is some video which I failed to find now. But better yet, there is [open source reference implementation right here on GitHub](https://github.com/vrk-kpa/digiid-eservice-integration-reference) with links to [guide for RP](https://wiki.dvv.fi/display/DHHJD/SIOPv2+POC+-+Guide+for+Relying+Parties). It should open the identity service for any services on Internet next year. Any passport or ID card holder can enable the wallet and can prove many attributes (unique ID, name, old enough, picture, address etc.) to services that accept them. There are plans to get drivers licenses, certified competencies, library cards, memberships and other certificates included in the wallet for use when needed. Employers and such could provide their own certified information via the wallet. Perhaps eventually even Mastodon could issue membership card to be stored in the wallet for whatever use. If the wallet supports many issuers, the membership cards will not need to be issued by other authorities.

If you search for digital identity on the web, you can find there are projects around the world. In Africa, India, Australia, etc. Others may use something else or just end up using the same standards becoming only new sources of trust.

I'm not only looking to be able to prove if I am a real person of age, but I also wish that if I lost my credentials or some cracker hijacked my account, I could go to authorities and get new digital identity to prove who I am and regain access to the service to kick out the criminal. If my ID was lost, authorities would revoke the old lost identity and issue a new. The unique ID would be the same on new identity and would then give me access to services I had been using under the old identity.

**Reply by Andre601** [reply]
From what I understand would this only be for indivduals, right? Not companies, groups, collectives, etc.

My discussion here is about a general verification thing, similar to Twitter's checkmark before Mr. Tesla-boss decided to sell it for a few bucks to everyone.

Like something that would show easily "This account belongs to a user/group/... that is real and not a faker/copy-cat!".
Link verification can only go so far and may not be a protection against fakers, scammers and alike in the end...

If that ID thing would work on companies and alike, I would look forward to it and see what it may hold in the end.

(On an unrelated note, I believe this ID thing may still suffer from the same issues that people with Plurality may have... If not, then that's a nice move forward into inclusion I suppose)

**Reply by kohtala** [reply]
This is the same level as you'd have if you asked people to come and identify themselves with a passport. Only that it works over Internet and people can choose what they give of themselves while on passport the inspector can see everything that is on it.

This is not yet available, but I thought you are interested in something that would be available in near future.

Purpose is to allow for businesses to identify themselves. I do not know details. Businesses could also adopt the technology to issue [verifiable credentials](https://www.w3.org/TR/2022/REC-vc-data-model-20220303/) to employees for example to establish the employment relation. If I follow the specs right (I've not worked on implementation), the issuer property DID value can be used to [fetch](https://w3c-ccg.github.io/did-method-web/) a DID document that carries [verificationMethod](https://www.w3.org/TR/did-core/#verification-methods) definition for the key. If RP trusts the employer as issuer, it can verify the credentials. There would be some use cases that could help adopt the technology, such as proving employment to security at the gate as well as access to IT systems, VPN etc. I did not find anyone had defined a common schema of attributes for employers to issue to employees.

If the above business identity does not exist or can not be trusted, direct trust could be created by the organization creating an account on Mastodon and giving their CA public key there. After that the credentials would be trusted to that organization. I find there is also https://coopcreds.com/ perhaps for developing some kind of co-operative trust model.

When there is digital business identity, technology would allow a person acting [on behalf of](https://www.w3.org/TR/2022/REC-vc-data-model-20220303/#holder-acts-on-behalf-of-the-subject) the organization, provided such verifiable credential has been issued and provided to RP.

I do not recognize what Plurality you refer to. The system does allow having many persona.

**Comment by DiggidyDave** [comment]
I've been thinking about this as well. It could be a bit less centralized if "verification provider services" (as an abstraction) are a pluggable feature by various servers. Then every server could be free to choose verification providers, and then there would need to only be some protocol level integration possibly that allowed the server-provided verification to follow the user around.

It might look a lot like certs on web pages, where you get an indicator that the "Connection is secure" and "this site has a valid certificate, issued by a trusted authority." But there are quite a few providers available for that cert. And there are a lot of pages that don't bother, which (wrt identity) might be an intentional decision/feature on some servers.

I'm newly interested in this, and I still need to take more time to deep dive into the architecture and protocols behind Mastodon, so take everything I say with a grain of salt, mostly I wanted to chime in to give support to this idea of pluggable services that can integrate with federated servers.

**Reply by jtheuerkauf** [reply]
Found this after your response to my comment.

i think we're alike on the idea of integrating identity providers. Ultimately "identity" has to have some trusted authority. Truly decentralized identity doesn't seem like it holds much water: "I am who I say I am because I say I am and don't need anyone else to confirm it. Trust me bro." There always has to be some authority to confirm the claim, which by definition becomes a centralized system.

Of course this is covered above, just chipping in my interpretation. i don't see viable options for fully decentralized verification. In my mind, there's no such thing. Setting that idea aside, multiple provider integrations or a single authority remain. i'm not sure i'm sold on the idea that instance moderators could choose which authorities they accept - imagine the havoc when other instances don't recognize the same authorities.

🤔

**Reply by mhus** [reply]
Verification should not be part of the server. It’s a client thing. - to have a second source of authority. The client should be able to choose the verification servers itself. And maybe a default one as proxy.

**Reply by DiggidyDave** [reply]
@jtheuerkauf now it's my turn to realize I missed your comment lol. I think SSL certificate authorities are a good analogy here. It isn't monopolized by a single source of truth, there are many providers to choose from, but they are each trusted authorities.

@mhus can you clarify what you mean by this? I'm having a hard time wrapping my head around how an unbounded number of individual, unrelated clients could each independently choose to have a different third party verify the identity of the author of something they are being served by a Mastodon instance. Actual identity verification requires a person or entity (ie the owner of an account) to work with the verification authority to "prove" they are who they say they are, it is intrinsically NOT a client thing. Am I missing something?

Going back to the cert authority analogy (actually it's barely an analogy; it's almost exactly the same thing), how would you implement "making SSL certs part of the browser" instead of server?

**Reply by DiggidyDave** [reply]
One awesome benefit of having pluggable id verification at the server level is that some servers would begin to carry trust with them throughout  the fediverse. eg "_Ah this user is on fosstidon.org which requires all users to verify with Digicert, I know I can trust these users and allow content from that server on my own Mastodon server because I only wanted a verified ecosystem._"

**Reply by Andre601** [reply]
Problem here is, that not everyone is well aware of ID verification services and especially how trustworthy they are.

It's kind of a balancing act I would say... Full freedom of chosing the verifier for your server may be nice, but can also be dangerous in the end because how can you be sure that "AwesomeVerifiedAccountsService" is trustworthy? On the other hand, only allowing a selected amount of options could be limiting in some ways, while trust could be better.

What I want to say is, that trust is something not everyone is on the same page with. One user trusts company X while the other user hates them and trusts Y instead.

**Reply by jtheuerkauf-neubus** [reply]
That's exactly the line of thinking i had. While there's no single authority to say i am who i say, the verification only works if the authority i choose is both recognized by other authorities and servers, and continues to tell them i'm one of its verified clients.

**Comment by swamidass** [comment]
I suppose you could charge $8 per month for it 😀

**Reply by Andre601** [reply]
Make it $5. Would be more attractive to people :^)

**Reply by swamidass** [reply]
Yes, good thinking. Let's undercut the blue bird =).

**Comment by f431** [comment]
Tim Pritlove mentioned in his Podcast [Logbuch:Netzpolitik](https://logbuch-netzpolitik.de/lnp445-das-goldkettchen-des-internets) (a German Podcast) he called "verified Bio". So to have a System to approve if this person is `this` and has `this` official function. 

So how could something like this be done. First it might be nice to distingue between `Person`, `Person of Influence/Interest`, `Bot` and `Organization` (Symbolized with e.g. a dot, circle, triangle, octagon next to the name). 

Depending what kind of Organization it is there might be different ways to determine what a Person, POI or Bot is, the rules what can be said and done is within the mastodon instance, but to verify that you are X or Y should be handled on a separate instance/system. 

A use case might be: A German politician in the European Parliament that hosts his/her own mastodon instance. There should probably be a DE-Verification System like @kohtala mentioned eIDAS that verifies that this mastodon handle is a German citizen. So there might be a hidden tag `NATIONALITY.DE` with a public key or your mastodon handle on a public DE-verification instance. Than there should be a EuroPar-Verification System that (probably manually) adds mastodon handles into a list that shows that he/she was in this function something like `POLFUNC.EU.Parliament.MEP.IX`. The "Admin" with the right to add should be handled within the Organization.

This politician might than write in his bio something like: "`@POLFUNC.EU.Parliament.MEP.IX` " that translates to "Member of the European Parliament" and has a blue/green color and a info text if this mastodon handle is in the EU-Verification System and red if not, and POI can show it within the mastodon handle via a circle left to it.   

So there might be a ROOT-Organization ideally something established like the UN, that authenticates suborganisations like Nations or to stay political neutral Partially Recognized Nations.

Priority might be a) Politicians and b) Press, but there might also be a use case for scientific data delivery e.g. if you have a weather station and toot every hour your measurements it might be nice if there is a Organization of weather forecasters to verify that your station is real and gives somehow accurate measurements (by physically inspecting it). 

There could also be a color scheme to see what general topic this person, bot or organization is associated with (e.g. white for religion, blue for politicians, black for journalists, green for activists, red for NSFW ...)

**Reply by Andre601** [reply]
Wouldn't that *kinda* caus the issue that one (or a few) big group(s) would "call the shots" here in terms of verification? At least that's how I actually understand it...

**Reply by kohtala** [reply]
Notice that I was thinking of the technology. EU within eIDAS is only one adopter, but the technology is not dependent on EU nor does it need any root organization. [Self-sovereign identity (SSI)](https://en.wikipedia.org/wiki/Self-sovereign_identity) allows people to carry verifiable credentials from many issuers in their digital identity wallet. What EU adds is credentials to prove age, name, nationality, etc., but employers or organizations could provide credentials to prove roles in their organizations.

To be practical, it needs service providers. I do not think all organizations are able to issue credentials without help from some service provider. I'm not familiar enough to know if there yet are any.

**Comment by Andre601** [comment]
I've updated my initial post to hopefully be a bit more structured... I also gave examples for the de-centralized solution I had in mind...

**Comment by Andre601** [comment]
I had time to think about a way and had an interesting (or stupid depending on how you see it) idea for how such a thing *could* work.
In the end would that depend on how performant Mastodon is in communicating with other instances of the Fediverse, as it would require communication between them on a frequent level (I assume).

The idea is relatively simple: Just imagine a voting system tbh.
A user is requesting verification on their instance. And when it goes through the first check, which is the admin approval, would the instance send info towards the connected instances, to inform about this new pending verification request.

Now every instance could give a Yes or No response to approve or deny the verification and once enough instances (~75% of all connected ones in this example) have given a vote, would the verification either be approved or denied. If the percentage of submited votes vs instances is not above the minimum or the votes resulted in an exact 50/50 split would the user remain in pending state. Users could only retry the approval after 30 days from their last attempt.

To avoid abuse, where an instance could be modifies and just claim a user is verified can the instances communicate with each other to collect the results and see if it makes sense. If not would it not be counted.

I think this is a good aproach. It would allow the system to stay Federated (Not a single entity chooses) and even makes it somewhat democratic by having the Fediverse vote on the approval.
I imagine the voting being based on each Instance's admins, but perhaps a public vote system, where each user of an instance can vote (With the same percentage of votes required to send a vote out) could also be an interesting aproach here?

Either way, that's my random idea I had... And for some reason did I call it "Fedivication" (Fediverse + Verification)... No real idea why.

**Comment by jtayler** [comment]
Mastodon automatically indicates “Verified” status for the TruAnon Identity.

This makes common social services easy to verify that are sometimes impossible using Mastodon alone. This naturally helps people build trust by sharing things about themselves that others can see and they can do this in way that is already commonly used to build trust on Mastodon profiles.

https://mastodon.social/@jtayler

![image](https://user-images.githubusercontent.com/1004984/202792318-98bbb7f3-c75f-433c-b6a1-0fcfa1a2c57c.png)

**Reply by Andre601** [reply]
While this looks like an interesting thing, are there (at least for me) still some problems...

The biggest is (yet again) that it wouldn't be federated. It would be one central place for the profile to be verified, which again gives control to one entity rather than multiple individuals (I base this of the info I gathered from the truanon site).

Next did I never hear of this service at all... And if I didn't hear of it, so did most casual people probably too, which means it's a relatively unknown source and (ironically) not that trustworthy to begin with. Of course, the site's goals can and probs are good, but there is always this level of info missing for something you never saw before, which plants a seed of doubt... I hope you get what I want to say with that.

**Reply by jtayler** [reply]
No, it just supports the same link confirmation system Mastodon already uses. You don't trust any authority - it leaves your identity untouched and in-place etc. etc. - https://jtayler.medium.com/free-and-easy-user-verification-on-mastodon-8566f239e2be

**Comment by jtheuerkauf** [comment]
Would this be the appropriate discussion to suggest verification with ID.me? i'd definitely like to see some other verification methods than having to set up a website just to add a meta tag to a page. i'm brand new, so maybe there are options i'm unaware of...

**Reply by Andre601** [reply]
The idea of this discussion was more about discussing possible ways of a decentralized verification system, meaning that not one site/group alone is responsible for it...

**Reply by jtheuerkauf** [reply]
Right, i wouldn't suggest all verification is handed to ID.me - only that it could be added to options for verification. But i guess even if it's _among_ several options, by itself it's considered centralized? i'm new to the concept - lots to learn. Thanks for the response.

**Reply by DiggidyDave** [reply]
Repeating something I said above, because I think it aligns with what @jtheuerkauf is saying (IIUC): an identity verification interface that allows providers to be "plugged in" would not centralize it, it would only provide some standards and provide a good user experience. Allowing users to choose and engage with an identity verification service that seamlessly integrates with mastodon would be a fantastic improvement over making them setup another webpage, etc, which is not intuitive and is a very poor experience.

Federation is not incompatible with having trusted, reputable entities plugged into things.

**Reply by jtheuerkauf** [reply]
Exactly what i meant. Thank you. 😀

**Comment by pixiekat** [comment]
When dealing with verification, I urge people to consider making it as easy as possible while not losing security. Most people who would benefit best from verification would be newsmakers, politicians, state officials, etc - assuming that they're technically savvy is something that shouldn't be done. If people get frustrated too easily because of this, they might just give up on Mastodon completely.

**Comment by roke-julian-lockhart** [comment]
https://github.com/mastodon/mastodon/issues/913
